src/combinatorics/young_diagram.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2022 Jake Levinson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jake Levinson
+-/
+import order.upper_lower
+
+/-!
+# Young diagrams
+
+A Young diagram is a finite set of up-left justified boxes:
+
+```text
+□□□□□
+□□□
+□□□
+□
+```
+This Young diagram corresponds to the [5, 3, 3, 1] partition of 12.
+
+We represent it as a lower set in `ℕ × ℕ` in the product partial order. We write `(i, j) ∈ μ`
+to say that `(i, j)` (in matrix coordinates) is in the Young diagram `μ`.
+
+## Main definitions
+
+- `young_diagram` : Young diagrams
+- `young_diagram.card` : the number of cells in a Young diagram (its *cardinality*)
+- `young_diagram.distrib_lattice` : a distributive lattice instance for Young diagrams
+  ordered by containment, with `(⊥ : young_diagram)` the empty diagram.
+
+## Notation
+
+In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). This terminology is used
+below, e.g. in `young_diagram.nw_of`.
+
+## Tags
+
+Young diagram
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+/-- A Young diagram is a finite collection of cells on the `ℕ × ℕ` grid such that whenever
+a cell is present, so are all the ones above and to the left of it. Like matrices, an `(i, j)` cell
+is a cell in row `i` and column `j`, where rows are enumerated downward and columns rightward.
+
+Young diagrams are modeled as finite sets in `ℕ × ℕ` that are lower sets with respect to the
+standard order on products. -/
+@[ext] structure young_diagram :=
+(cells : finset (ℕ × ℕ))
+(is_lower_set : is_lower_set (cells : set (ℕ × ℕ)))
+
+namespace young_diagram
+
+instance : set_like young_diagram (ℕ × ℕ) :=
+{ coe            := coe young_diagram.cells,
+  coe_injective' := λ μ ν h, by { rwa [young_diagram.ext_iff, ← finset.coe_inj] } }
+
+@[simp] lemma mem_cells {μ : young_diagram} (c : ℕ × ℕ) :
+  c ∈ μ.cells ↔ c ∈ μ := iff.rfl
+
+/-- In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+    means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). -/
+lemma nw_of (μ : young_diagram) {i1 i2 j1 j2 : ℕ}
+  (hi : i1 ≤ i2) (hj : j1 ≤ j2) (hcell : (i2, j2) ∈ μ) : (i1, j1) ∈ μ :=
+μ.is_lower_set (prod.mk_le_mk.mpr ⟨hi, hj⟩) hcell
+
+section distrib_lattice
+
+instance : has_le young_diagram :=
+{ le := λ μ ν, (μ : set (ℕ × ℕ)) ⊆ ν }
```

Comment:
There are some missing lemmas here about `μ.cells ⊆ ν.cells` and `(μ : set (ℕ × ℕ)) ⊆ ν`; and similarly for `<` / `\ssubset`.

Reply:
@eric-wieser Are these missing in the "would be nice to have to fill out the API" or the "they must be included in the same PR that introduces a `has_le` instance" sense?

If it's the latter, I think it would help keep the PR process going if you or someone contributed code that will pass review -- that way Jake doesn't have to guess what you mean and then be required to go through more rounds of review.

Reply:
I intend it in the latter sense.

I've pushed the appropriate change; in fact this `has_le` instance is redundant, as we get it for free via `set_like`. 
I've removed it.

Reply:
Great, thanks @eric-wieser !

Reply:
Thanks @eric-wieser ! And thanks @kmill, I had been thinking I indeed don't know how to tell when there's more stuff that needs to be added for the PR to be complete. Mind you, if these sorts of requirements are written up clearly somewhere, I'm happy to learn them.